### ISSUE TITLE
c++ support 

### ISSUE THREAD

**Issue by sidsarasvati** [awslabs/aws-sam-cli#914]
### Describe your idea/feature/enhancement

Provide a clear description. Ex. I wish SAM CLI would [...]

### Proposal

Add details of how to add this to the product.

Things to consider:
1. Will this require any updates to the [SAM Spec](https://github.com/awslabs/serverless-application-model)

**Comment by jfuss** [comment]
@sidsarasvati Can you be more descriptive what you mean by support? C++ is supported through the provided runtime.

**Comment by jfuss** [comment]
C++ is supported through the provided runtime, which is the same as the AWS Lambda Service.

If something else was meant by support, please open a new issue and clearly describe the feature or enhancement.

**Comment by garymazz** [comment]
I think this issue warrants consideration. It would be beneficial to provide an example and documentation substantiating a claim of support .  Whether the issue should be considered part of aws-sam-cli or serverless-application-model projects is an issue best worked out between project leads and contributors.

I'm also looking for a way to launch Lambda's c++ motor (runtime) via SAM. More specifically, information indicating whether it's supported using "sam local".  However 6 months after the announcement, there is no 'direct' example or documentation on launching the c++ runtime using SAM CLI; users are left wondering.

@sidsarasvati I'm just building the runtime and this has not been verified, but to answer  your question, command line: "sam local start-api" and "sam local invoke xxxx" is likely the best candidates.  Whether the SAM motor supports the c++ runtime is a question.

@jfuss, I think if you could add a bit more detail to your explanation, it could beneficial to others unfamiliar with the technology. 
b/r

**Comment by jfuss** [comment]
@garymazz Using C++ in SAM CLI is (or should be) no different than using it on the Lambda Service. That is, the C++ runtime is a provided (custom) runtime. You will need to add the layer to your function definition ([SAM](https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#awsserverlessfunction) or [CloudFormation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-layers)). You can read our [docs](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-layers.html) for more details on Layers as well.

**Comment by garymazz** [comment]
@jfuzz.. Thanks for the references :) 

I didn't think there was a difference between platforms, at least from the cli perspective.  In the aws world, there's seems to be an assumption that everything sort of works together. Developers coming from the open source world assume if something is 'not' documented, it likely doesn't work. Also, the opposite is true, things claim to work but don't.  For example, attempt building xerces-c on cygwin, the documentation claims it works using autotools, but it doesn't. 

I'll be spending this week, time permitting, creating a cling/clang custom runtime based off the c++ runtime image.  I do wish there was a linux docker image released as part of the aws-sam-cli distribution in-lieu having atp and yum presence. A linux docker image would save a bunch of time (removing a barrier to engagement)  sorting through brew (not a fan) and pip install dependency issues (on centos7).  I do understand the reasons for adopting brew, this is not a request to abandon the tools chain. 

Thanks again :)

**Comment by daniel-fudge** [comment]
@jfuss Thanks for the positive sounding note. 
I've read these documents and many others but I've yet to find a simple example of a C++ Lambda function deployed with SAM.
The following are well documented and I have performed them successfully.
1. Python Lambda functions with layers in SAM
2. Build C++ lambda functions with the SDK
3. Deploy C++ Lambda function on the CLI 

I also see the Rust and Python example here.
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/building-custom-runtimes.html

I think many users would benefit from a simple example such as the following but deployed with SAM.
https://github.com/awslabs/aws-lambda-cpp

Basically provide the SAM equivalent of:
```shell
$ aws lambda create-function --function-name demo \
--role <specify role arn from previous step here> \
--runtime provided --timeout 15 --memory-size 128 \
--handler demo --zip-file fileb://demo.zip
```
